sound/soc/codecs/max98373-sdw.c

```diff
@@ -0,0 +1,1203 @@
+{
+	struct max98373_sdw_priv *max98373 = dev_get_drvdata(dev);
+
+	regcache_cache_only(max98373->regmap, false);
+	regcache_sync(max98373->regmap);
+	return 0;
+}
+
+static const struct dev_pm_ops max98373_pm = {
+	SET_SYSTEM_SLEEP_PM_OPS(max98373_suspend, max98373_resume)
+};
+/*                                                                           */
+/*****************************************************************************/
+
+
+/*****************************************************************************/
+/*     MAX98373 Reset and initialization.                                    */
+
+static void max98373_reset(struct sdw_slave *slave)
+{
+	int ret, reg, count;
+	struct device *dev = &slave->dev;
+	struct max98373_sdw_priv *max98373 = dev_get_drvdata(dev);
+
+	/* Software Reset */
+	ret = regmap_update_bits(max98373->regmap,
+		MAX98373_R2000_SW_RESET,
+		MAX98373_SOFT_RESET,
+		MAX98373_SOFT_RESET);
+	if (ret)
+		dev_err(dev, "Reset command failed. (ret:%d)\n", ret);
+
+	count = 0;
+	while (count < 3) {
+		usleep_range(10000, 11000);
+		/* Software Reset Verification */
+		ret = regmap_read(max98373->regmap,
+			MAX98373_R21FF_REV_ID, &reg);
+		if (!ret) {
+			dev_info(dev, "Reset completed (retry:%d)\n", count);
+			return;
+		}
+		count++;
+	}
+	dev_err(dev, "Reset failed. (ret:%d)\n", ret);
+}
+
+static int max98373_io_init(struct sdw_slave *slave)
+{
+	struct device *dev = &slave->dev;
+	struct max98373_sdw_priv *max98373 = dev_get_drvdata(dev);
+
+	/* Perform IO operations only if slave is in ATTACHED state */
+	if (slave->status != SDW_SLAVE_ATTACHED)
+		return 0;
+
+	/* Enable Runtime PM */
+	pm_runtime_set_autosuspend_delay(dev, 3000);
+	pm_runtime_use_autosuspend(dev);
+	pm_runtime_enable(dev);
+
+	/* Software Reset */
+	max98373_reset(slave);
+
+	regmap_write(max98373->regmap, MAX98373_R20FF_GLOBAL_SHDN, 0);
+	/* Set soundwire mode */
+	regmap_write(max98373->regmap, MAX98373_R2025_AUDIO_IF_MODE, 3);
+	/* Enable ADC */
+	regmap_write(max98373->regmap, MAX98373_R2047_IV_SENSE_ADC_EN, 3);
+	/* Set default speaker volume */
+	regmap_write(max98373->regmap, MAX98373_R203D_AMP_DIG_VOL_CTRL, 0);
+	/* Enable speaker */
+	regmap_write(max98373->regmap, MAX98373_R2043_AMP_EN, 3);
+
+	/* Set default Soundwire clock */
+	regmap_write(max98373->regmap, MAX98373_R2036_SOUNDWIRE_CTRL, 5);
+	/* Set default sampling rate for speaker and IVDAC */
+	regmap_write(max98373->regmap, MAX98373_R2028_PCM_SR_SETUP_2, 0x88);
+	/* Amp enable */
+	regmap_write(max98373->regmap, MAX98373_R20FF_GLOBAL_SHDN, 1);
+
+	pm_runtime_put_sync_autosuspend(dev);
+
+	return 0;
+}
+
+
+static int max98373_snd_probe(struct snd_soc_component *component)
+{
+	struct max98373_sdw_priv *max98373 =
+		snd_soc_component_get_drvdata(component);
+
+	regmap_write(max98373->regmap, MAX98373_R20FF_GLOBAL_SHDN, 0);
+
+	/* Enable the codec - it can play the audio after 4 ms....	*/
+	regmap_write(max98373->regmap, MAX98373_R20FF_GLOBAL_SHDN, 1);
```

Comment:
Sorry, I don't quite understand this comment. I thought regmap is usable once 'devm_regmap_init_sdw' is done. Why do we need to check codec enumeration?
'max98373_snd_probe' function is called after codec is registered.
These superfluous SHDN control has been removed from probe function.

Reply:
the sequence is as follows:

1. the bus walks through the list of ACPI devices, and will send a modalias to userspace to load the relevant driver.
2. the module is inserted, which results in the driver being registered and you probe the codec driver.
3. the bus starts and exit resets
4. the codec reports its presence and gets enumerated as non-zero Device
5. only now can the transactions with the device start. You  *CANNOT* start a transaction, regmap or otherwise, before enumeration is complete. That's precisely what we added these 'struct completion' utilities so that codec drivers know when it is safe to transfer data on the bus.

Reply:
Thanks for the explanation. Is the enumeration process only logical process without actual h/w communication?
I see most sdw driver check 'initialization_complete' in resume function only. Do you think this need to be done in amp probe function as well?

Reply:
no, enumeration does entail reading DeviceID registers, and writing a non-zero device number to the hardware. But that is done in a 'controlled' way, and only when the device is known to be attached on the bus.
The probe should only do basic things, such as allocation of data, then when the hardware is detected you will get a callback. At that point you can talk to the device and sync all the regmap stuff. see the io_init() routines in rt*sdw.c drivers.

Reply:
That's great to know. Thank you. I confirmed all regmap access before io_init have been removed.